domain/modelagent/state/modelstate.go

```diff
@@ -387,6 +513,139 @@ WHERE machine_uuid = $machineUUIDRef.machine_uuid
+          a.name AS &unitAgentBinaryMetadata.architecture_name,
+          osm.size AS &unitAgentBinaryMetadata.size,
+          osm.sha_256 AS &unitAgentBinaryMetadata.sha_256,
+          osm.sha_384 AS &unitAgentBinaryMetadata.sha_384
+FROM      unit_agent_version AS uav
+JOIN      unit AS u ON uav.unit_uuid = u.uuid
+JOIN      architecture AS a ON uav.architecture_id = a.id
+LEFT JOIN v_agent_binary_store AS abs ON (
+          uav.version = abs.version
+AND       uav.architecture_id = abs.architecture_id)
+LEFT JOIN object_store_metadata AS osm ON abs.object_store_uuid = osm.uuid
+`, unitAgentBinaryMetadata{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	unitCount := rowCount{}
+	stmtUnitCount, err := st.Prepare(`
+SELECT (count(*)) AS (&rowCount.count)
+FROM   unit
+`, unitCount)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	unitBinaryMetadata := []unitAgentBinaryMetadata{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt).GetAll(&unitBinaryMetadata)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf(
+				"getting unit binary information from database: %w", err,
+			)
+		}
+
+		if err := tx.Query(ctx, stmtUnitCount).Get(&unitCount); err != nil {
+			return errors.Errorf(
+				"getting the number of units currently in the model: %w", err,
+			)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	if len(unitBinaryMetadata) != unitCount.Count {
+		return nil, errors.New(
+			"not all units in the model have their agent version set",
+		).Add(modelagenterrors.UnitAgentVersionNotSet)
+	}
+
+	rval := make(map[coreunit.Name]coreagentbinary.Metadata, len(unitBinaryMetadata))
+	for _, unitRecord := range unitBinaryMetadata {
+		// Because we are performing a left join against agent binary store with
+		// no RI there exists the possibility that unit might be using a
+		// version that isn't in the model store. In theory this should never
+		// happen. We need to check all of the agent binary store values from
+		// the query for null to work out if this case exists.
+		//
+		// In theory just checking one of these values should be enough to
+		// identify the condition but all three are done for safety and
+		// correctness.
+		if !unitRecord.SHA256.Valid ||
+			!unitRecord.SHA384.Valid ||
+			!unitRecord.Size.Valid {
+			return nil, errors.Errorf(
+				"unit %q has missing agent binaries in the model",
+				unitRecord.UnitName,
+			).Add(modelagenterrors.MissingAgentBinaries)
+		}
+
+		number, err := semversion.Parse(unitRecord.Version)
+		if err != nil {
+			return nil, errors.Errorf(
+				"parsing unit %q version %q number: %w",
+				unitRecord.UnitName, unitRecord.Version, err,
+			)
+		}
+
+		unitName, err := coreunit.NewName(unitRecord.UnitName)
+		if err != nil {
+			return nil, errors.Errorf(
+				"parsing unit name %q: %w", unitRecord.UnitName, err,
+			)
+		}
```

Comment:
We can just cast `unitRecord.UnitName` to `coreunit.Name` since the unit name is validated on the way in.